nifi-commons/nifi-security-utils/src/test/java/org/apache/nifi/security/util/crypto/CipherUtilityTest.java

```diff
@@ -0,0 +1,322 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class CipherUtilityTest {
+    private static final Pattern KEY_LENGTH_PATTERN = Pattern.compile("([\\d]+)BIT");
+    // TripleDES must precede DES for automatic grouping precedence
+    private static final List<String> CIPHERS = Arrays.asList("AES", "TRIPLEDES", "DES", "RC2", "RC4", "RC5", "TWOFISH");
+    private static final List<String> SYMMETRIC_ALGORITHMS = Arrays.stream(EncryptionMethod.values())
+            .map(it -> it.getAlgorithm())
+            .filter(algorithm -> algorithm.startsWith("PBE") || algorithm.startsWith("AES"))
+            .collect(Collectors.toList());
+    private static final Map<String, List<String>> ALGORITHMS_MAPPED_BY_CIPHER = SYMMETRIC_ALGORITHMS
+            .stream()
+            .collect(Collectors.groupingBy(algorithm -> CIPHERS.stream().filter(cipher -> algorithm.contains(cipher)).findFirst().get()));
+
+    // Manually mapped as of 03/21/21 1.13.0
+    private static final Map<Integer, List<String>> ALGORITHMS_MAPPED_BY_KEY_LENGTH = new HashMap<>();
+    static {
+        ALGORITHMS_MAPPED_BY_KEY_LENGTH.put(40, Arrays.asList("PBEWITHSHAAND40BITRC2-CBC",
+                "PBEWITHSHAAND40BITRC4"));
+        ALGORITHMS_MAPPED_BY_KEY_LENGTH.put(64, Arrays.asList("PBEWITHMD5ANDDES",
+                "PBEWITHSHA1ANDDES"));
+        ALGORITHMS_MAPPED_BY_KEY_LENGTH.put(112, Arrays.asList("PBEWITHSHAAND2-KEYTRIPLEDES-CBC",
+                "PBEWITHSHAAND3-KEYTRIPLEDES-CBC"));
+        ALGORITHMS_MAPPED_BY_KEY_LENGTH.put(128, Arrays.asList("PBEWITHMD5AND128BITAES-CBC-OPENSSL",
+                "PBEWITHMD5ANDRC2",
+                "PBEWITHSHA1ANDRC2",
+                "PBEWITHSHA256AND128BITAES-CBC-BC",
+                "PBEWITHSHAAND128BITAES-CBC-BC",
+                "PBEWITHSHAAND128BITRC2-CBC",
+                "PBEWITHSHAAND128BITRC4",
+                "PBEWITHSHAANDTWOFISH-CBC",
+                "AES/CBC/NoPadding",
+                "AES/CBC/PKCS7Padding",
+                "AES/CTR/NoPadding",
+                "AES/GCM/NoPadding"));
+        ALGORITHMS_MAPPED_BY_KEY_LENGTH.put(192, Arrays.asList("PBEWITHMD5AND192BITAES-CBC-OPENSSL",
+                "PBEWITHSHA256AND192BITAES-CBC-BC",
+                "PBEWITHSHAAND192BITAES-CBC-BC",
+                "AES/CBC/NoPadding",
+                "AES/CBC/PKCS7Padding",
+                "AES/CTR/NoPadding",
+                "AES/GCM/NoPadding"));
+        ALGORITHMS_MAPPED_BY_KEY_LENGTH.put(256, Arrays.asList("PBEWITHMD5AND256BITAES-CBC-OPENSSL",
+                "PBEWITHSHA256AND256BITAES-CBC-BC",
+                "PBEWITHSHAAND256BITAES-CBC-BC",
+                "AES/CBC/NoPadding",
+                "AES/CBC/PKCS7Padding",
+                "AES/CTR/NoPadding",
+                "AES/GCM/NoPadding"));
+    }
+
+    @BeforeAll
+    static void setUpOnce() {
+        Security.addProvider(new BouncyCastleProvider());
```

Comment:
In this case, the security provider registration is necessary, so this line can be retained, while others should be removed.

Reply:
Will check the tests one by one